### PR TITLE
(PA-274) add sles-11-s390x platform

### PR DESCRIPTION
**Pull request by ScottGarman**
We cross-compile this platform on a sles-11-x86_64 pooler system.

It blends! The only work left to do is to include puppet-ca-bundle, which needs some work in a cross-compiled environment.

### CODE REVIEW

File: configs/components/curl.rb [hunk 1]
```diff
@@ -4,7 +4,12 @@
   pkg.url "http://buildsources.delivery.puppetlabs.net/curl-#{pkg.get_version}.tar.gz"
 
   pkg.build_requires "openssl"
-  pkg.build_requires "puppet-ca-bundle"
+  pkg.build_requires "puppet-ca-bundle" unless platform.architecture == "s390x"
```
Comment (stahnma) [inline configs/components/curl.rb]: Can you put a comment here about how this is currently a known issue and should be fixed?

File: configs/components/facter.rb [hunk 2]
```diff
@@ -112,7 +112,7 @@
     skip_blkid = 'OFF'
   elsif platform.is_rpm?
     if (platform.is_el? && platform.os_version.to_i >= 6) || (platform.is_sles? && platform.os_version.to_i >= 11) || platform.is_fedora?
-      pkg.build_requires "libblkid-devel"
+      pkg.build_requires "libblkid-devel" unless platform.architecture == "s390x"
```
Comment (stahnma) [inline configs/components/facter.rb]: I *think* rubocop wants || vs or

File: configs/components/virt-what.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@
     requires "util-linux"
   end
   if platform.name =~ /^sles-(10|11)-.*$/
-    requires "pmtools"
+    requires "pmtools" unless platform.architecture == "s390x"
```
Comment (stahnma) [inline configs/components/virt-what.rb]: Why is it not included on s390?

File: configs/platforms/sles-11-s390x.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+platform "sles-11-s390x" do |plat|
```
Comment (stahnma) [inline configs/platforms/sles-11-s390x.rb]: Maybe a comment at the top that explains this is cross compiled.

File: resources/files/rbconfig-s390x-linux-gnu.rb [hunk 5]
```diff
@@ -0,0 +1,268 @@
+
+# This file was created by PuppetAgent for HuaweiOS (Debian 8 PPC).  Any
```
Comment (stahnma) [inline resources/files/rbconfig-s390x-linux-gnu.rb]: This seems incorrect for SLES s390x

File: resources/files/rbconfig-s390x-linux-gnu.rb [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+# changes made to this file will be lost the next time ruby is built.
```
Comment (stahnma) [inline resources/files/rbconfig-s390x-linux-gnu.rb]: This file is generated by building Ruby on the platform (minus the cross compile, have we done that?)

File: configs/components/facter.rb (hunk 2 above)
Comment (ScottGarman) [inline configs/components/facter.rb]: rubocop isn't complaining about this. From what I've just looked up, the differences between using "or" vs. "||" only come into play when doing things like using them in assignment, and as if conditionals they are synonymous. Maybe @mckern can weigh in here?

Comment (mckern) [inline configs/components/facter.rb]: `||` and `or` is basic precedence; how will the conditional be evaluated from left to right? But as a rule, always use `||`. It's consistent and predictable.

**Comment by ScottGarman** [comment]
I've updated the PR with all feedback so far. Just to be paranoid-level safe, I'm running it through jenkins:

https://jenkins.puppetlabs.com/view/puppet-agent/view/ad%20hoc/job/platform_puppet-agent_init-van-int_suite-manual-ad-hoc/121/

File: configs/components/augeas.rb [hunk 7]
```diff
@@ -37,7 +39,11 @@
       pkg.build_requires 'runtime'
     end
 
-    pkg.build_requires 'pkgconfig'
+    if platform.architecture == "s390x"
+      pkg.environment "PATH" => "/opt/pl-build-tools/bin:$$PATH:#{settings[:bindir]}"
+      pkg.environment "CFLAGS" => settings[:cflags]
+      pkg.environment "LDFLAGS" => settings[:ldflags]
```
Comment (mckern) [inline configs/components/augeas.rb]: Probably out of scope for this PR, but why don't we set `:CFLAGS` and `:LDFLAGS` for everything everywhere all the time forever?

File: configs/components/curl.rb [hunk 8]
```diff
@@ -4,7 +4,14 @@
   pkg.url "http://buildsources.delivery.puppetlabs.net/curl-#{pkg.get_version}.tar.gz"
 
   pkg.build_requires "openssl"
-  pkg.build_requires "puppet-ca-bundle"
+  # TODO: remove the exception here once puppet-ca-bundle can be
+  # installed in cross-compiled environments
```
Comment (mckern) [inline configs/components/curl.rb]: We can fix this. Not in this PR but we can fix this. Can you please sign/date this comment?

File: configs/components/augeas.rb (hunk 7 above)
Comment (ScottGarman) [inline configs/components/augeas.rb]: Totally agree, this is something to discuss next week when we have a work session to talk about refactoring the p-a config files, where I'll be dumping some ideas out of my head that have been rattling around for a while.

**Comment by ScottGarman** [comment]
Ok, final tweaks made, and I fixed the silly md5 issue on the rbconfig file that caused the jenkins build to fail for sles-11-s390x, and rebuilt it locally to confirm it works. Ready for merge now.

File: configs/platforms/sles-11-s390x.rb (hunk 4 above)
Comment (mckern) [inline configs/platforms/sles-11-s390x.rb]: Seconded on cross-compile comment.

Comment (ScottGarman) [inline configs/platforms/sles-11-s390x.rb]: I followed the example I used for huaweios and the comment is just above the line that sets vmpooler_template.

Comment (mckern) [inline configs/platforms/sles-11-s390x.rb]: Dig it

**Comment by mckern** [comment]
:+1: 
I see things we can polish, smooth over, or remove but they're deeper issues that we should in the P-A codebase. Great job.

File: configs/components/leatherman.rb [hunk 9]
```diff
@@ -94,7 +94,7 @@
 
   # Make test will explode horribly in a cross-compile situation
   # Tests will be skipped on AIX until they are expected to pass
-  if platform.architecture == 'sparc' || platform.is_aix? || platform.is_huaweios?
+  if platform.architecture == 'sparc' || platform.architecture == 's390x' || platform.is_aix? || platform.is_huaweios?
```
Comment (shermdog) [inline configs/components/leatherman.rb]: Did we ever add an is_cross_compile flag?

Comment (ScottGarman) [inline configs/components/leatherman.rb]: No. Once this gets merged I intend to start working on that. Then the refactoring can begin.

**Pull request merged**